airflow/jobs/scheduler_job.py

```diff
@@ -1705,62 +1305,216 @@ def _run_scheduler_loop(self) -> None:
+            DagRun.external_trigger.is_(False),
+        ).scalar()
+
+        # TODO[HA]: add back in dagrun.timeout
+
+        if dag.max_active_runs and active_runs_of_dag >= dag.max_active_runs:
+            self.log.info(
+                "DAG %s is at (or above) max_active_runs (%d of %d), not creating any more runs",
+                dag.dag_id, active_runs_of_dag, dag.max_active_runs
+            )
+            dag_model.next_dagrun = None
+            dag_model.next_dagrun_create_after = None
+        else:
+            next_dagrun_info = dag.next_dagrun_info(next_run_date)
+            if next_dagrun_info:
+                dag_model.next_dagrun = next_dagrun_info['execution_date']
+                dag_model.next_dagrun_create_after = next_dagrun_info['can_be_created_after']
+            else:
+                dag_model.next_dagrun = None
+                dag_model.next_dagrun_create_after = None
+
+        # TODO[HA]: Should we do a session.flush() so we don't have to keep lots of state/object in
+        # memory for larger dags? or expunge_all()
+
+    def _schedule_dag_run(self, dag_run: DagRun, session: Session) -> int:
+        """
+        Make scheduling decisions about an individual dag run
+
+        :return: Number of tasks scheduled
+        """
+        dag_run.dag = self.dagbag.get_dag(dag_run.dag_id, session=session)
+
+        if not dag_run.dag:
+            self.log.error(
+                "Couldn't find dag %s in DagBag/DB!", dag_run.dag_id
+            )
+            return 0
+
+        if dag_run.execution_date > timezone.utcnow() and not dag_run.dag.allow_future_exec_dates:
+            self.log.error(
+                "Execution date is in future: %s",
+                dag_run.execution_date
+            )
+            return 0
+
+        # TODO: This query is probably horribly inefficient (though there is an
+        # index on (dag_id,state)). It is to deal with the case when a user
+        # clears more than max_active_runs older tasks -- we don't want the
+        # scheduler to suddenly go and start running tasks from all of the
+        # runs. (AIRFLOW-137/GH #1442)
+        #
+        # The longer term fix would be to have `clear` do this, and put DagRuns
+        # in to the queued state, then take DRs out of queued before creating
+        # any new ones
+        if dag_run.dag.max_active_runs:
+            currently_active_runs = session.query(func.count(TI.execution_date.distinct())).filter(
+                TI.dag_id == dag_run.dag_id,
+                TI.state.notin_(State.finished())
+            ).scalar()
+
+            if currently_active_runs >= dag_run.dag.max_active_runs:
+                return 0
+
+        # TODO[HA]: Run verify_integrity, but only if the serialized_dag has changed
+
+        # TODO[HA]: Rename update_state -> schedule_dag_run, ?? something else?
+        schedulable_tis = dag_run.update_state(session=session)
```

Comment:
This is where I'm concerned the most. `update_state` in our use case is very heavy as it requires task dependency checks. In our case there are extreme size DAGs that the dependency checks take minutes--the biggest DAG (20k unique tasks in it) takes >10 mins to check the dependencies for its 3 active DAG runs. It was problematic but in the old world it blocks task scheduling for those DAGs only but if we move it here it will block all task scheduling. Do you mind also elaborate a bit about the reason we want to extract this part from DAG processors and put it in the scheduling loop please?

Reply:
I'll respond fuller later, and run tests with a huge dag like you talk about, but I wonder if https://github.com/apache/airflow/pull/4751 has already massively improved that problem.

The cliff notes version: one of the major bottlenecks when having 100s or 1000s of dags was waiting for the parsing process to come around to the file again before the next task in a day could be scheduled.

Additional hitting the dag processor timeout (introduced/actually enforced in 1.10.4) could lead to a situation where a dag becomes permanently unschedulable: if it timesout (usually due to accumulating load/too many active dag runs, etc.) then that dag is never returned to the scheduler, and tasks in that day will never be eligible to be queued in the previous implementation - this has been surprising and hidden/hard to debug problem for many users causing production outages.

Reply:
And, to be clear, if the performance isn't good enough I'll change something beck to make in non-blocking in some way

Reply:
Out of interest how does the browser cope with displaying such a large dag (ignoring webserver boot up time)

Reply:
I have some ideas for further optimization of update_state - namely there's no point checking the ti deps of all 20k tasks every time - I think with a bit of thought (taking care of trigger rules etc) we could "prune" whole parts of the graph and not have to check them when the upstream tasks haven't run.

Are you able to share the "shape" of your large dag? (Tasks, triggers etc ). That would help - even if you can do it privately and I can turn it in to a representative test case

Reply:
Very interesting discussion. I did not have time to pursue it in detail but I will keep an eye on this and see if I can dig a bit deeper here .

Reply:
@KevinYang21 -> do you think we need more elaboration in this part? Or are you ok with this now?

Reply:
I still have some more testing to do here, but between https://github.com/apache/airflow/pull/4751 and https://github.com/apache/airflow/pull/11147 this part of the code is a lot quicker than early 1.10.x that Kevin is based off.

Reply:
Finally found time to do some testing, on my local laptop running against this PR, here is the code I've been testing with:

```
    for count in range(repeat):
        with db.create_session() as session:
            dagrun = session.query(DR).filter(DR.dag_id == dag.dag_id).first()
            dagrun.dag = dag
            start = time.perf_counter()

            dagrun.update_state(session=session)
            times.append(time.perf_counter() - start)
        print("Run %d time: %.5f" % (count + 1, times[-1]))

        if count + 1 != repeat:
            with db.create_session() as session:
                reset_dag(dag, session)

            create_dag_runs(dag, num_runs, session)

```

(`reset_dag` deletes everything from TI and DR tables, `create_dag_runs` does what it says, creates 1 (or more) runs, just to add more rows to the TI table.)

And the numbers I got were:

```
Run 1 time: 2.80744
Run 2 time: 3.07741
Run 3 time: 6.30900
Run 4 time: 6.34341
Run 5 time: 6.56565


Time for 5 dag runs of 1 dags with 21407 total tasks: 5.0206s (±1.902s)
```

Before #11147 it was massively _much_ more variable:

```
Run 1 time: 11.21811
Run 2 time: 142.50100
Run 3 time: 16.76023
Run 4 time: 151.12660
Run 5 time: 161.86329


Time for 5 dag runs of 1 dags with 21407 total tasks: 96.6938s (±75.835s)
```

And on 1.10.12 (with a bit of a change in the benchmark code to check the TI deps too (it turns out udpate_state didn't used to do schedule TIs -- that was only added in #4751) is... slow. I'll update this comment with timing when it actually finishes.7 minutes so far and it hasn't completed a single iteration.

@KevinYang21 I think this makes this approach _more_ than fast enough!

Reply:
@ashb I did try to cherry pick this pr to our internal repo,, which is based on 1.10.4: https://github.com/apache/airflow/pull/4751, (https://github.com/apache/airflow/pull/11010 for 1.10 branch) and https://github.com/apache/airflow/pull/7503. However, the overall performance actually increased. I have not found the root cause. It could be due to our optimization at the deps check with `ThreadPool`, which made if faster to do the deps check.

```
# this loop is quite slow as it uses are_dependencies_met for
# every task (in ti.is_runnable). This is also called in
# update_state above which has already checked these tasks
@provide_session
def enqueue_task_if_deps_met(ti, session=None):
    task = dag.get_task(ti.task_id)

    # fixme: ti.task is transient but needs to be set
    ti.task = task

    if ti.are_dependencies_met(
            dep_context=DepContext(flag_upstream_failed=True),
            session=session, early_exit_on_failed_dep=True):
        self.log.debug('Queuing task: %s', ti)
        return ti.key
    return None

with ThreadPoolExecutor(max_workers=50) as executor:
    for ti_key in filter(lambda x: x is not None, executor.map(enqueue_task_if_deps_met, tis)):
        task_instances_list.append(ti_key)

```

I think in general we should move to non blocking fashion to increase the throughput of the scheduler.

Reply:
These are some numbers after we cherry picked changes of #4751, (#11010 for 1.10 branch) and #7503

### before AIRFLOW-3607: Process a new dag run

```
[** REMOTE di-airflow-bronze-scheduler **]ping_zhang@i-0497283d24627a6ce:~/test-AIRFLOW-3607
$ tail -f log.txt
2020-10-09 21:52:18,144  Load DagBag: 87.1108751297 seconds
2020-10-09 21:52:42,324 Examining DAG run <DagRun minerva_compute @ 2020-10-09 21:52:18.238537+00:00: scheduled__2020-10-09T21:52:18.238537+00:00, externally triggered: False>
2020-10-09 21:54:20,709 Verify Integrity in 98.3845481873 seconds
2020-10-09 21:54:24,393 Update State in 3.68303990364 seconds
2020-10-09 21:56:34,579 Deps Check in 126.738894939 seconds
2020-10-09 21:56:34,579 process_tasks in 232.254889965 seconds with ready tis: 956
```

### after AIRFLOW-3607: : Process a new dag run 

```
2020-10-09 22:52:29,656 Load DagBag: 86.3233180046 seconds
2020-10-09 22:52:54,967 Examining DAG run <DagRun minerva_compute @ 2020-10-09 22:52:29.742831+00:00: scheduled__2020-10-09T22:52:29.742831+00:00, externally triggered: False>
2020-10-09 22:54:33,285 Verify Integrity in 98.3177959919 seconds
2020-10-09 22:56:27,847 Update State in 114.560956001 seconds
2020-10-09 22:56:27,854 process_tasks in 212.88758111 seconds ready tis: 956
```
----

### before AIRFLOW-3607: Process an existing  dag run with half root tasks in `success` state
```
2020-10-09 22:38:29,420 Load DagBag: 86.8423230648 seconds
2020-10-09 22:43:13,357 Examining DAG run <DagRun minerva_compute @ 2020-08-28 00:00:00+00:00: scheduled__2020-08-28T00:00:00+00:00, externally triggered: False>
2020-10-09 22:43:25,453 Verify Integrity in 12.0952570438 seconds
2020-10-09 22:43:29,027 Update State in 3.57410883904 seconds
2020-10-09 22:44:53,102 Deps Check perf in 83.3023910522 seconds
2020-10-09 22:44:53,102 process_tasks in 99.7442808151 seconds ready tis: 651
```

### after AIRFLOW-3607: Process an existing dag run with half root tasks in `success` state
```
2020-10-09 23:01:33,544 Load DagBag: 88.0396211147 seconds
2020-10-09 23:06:30,308 Examining DAG run <DagRun minerva_compute @ 2020-08-28 00:00:00+00:00: scheduled__2020-08-28T00:00:00+00:00, externally triggered: False>
2020-10-09 23:06:42,825 Verify Integrity in 12.516479969 seconds
2020-10-09 23:08:00,063 Update State in 77.2370829582 seconds
2020-10-09 23:08:00,066 process_tasks in 89.7578499317 seconds ready tis: 651
```